spring-context/src/main/java/org/springframework/scheduling/annotation/ScheduledAnnotationReactiveSupport.java

```diff
@@ -0,0 +1,209 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.scheduling.annotation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.time.Duration;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.reactivestreams.Publisher;
+import reactor.core.Disposable;
+import reactor.core.Disposables;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.aop.support.AopUtils;
+import org.springframework.core.CoroutinesUtils;
+import org.springframework.core.KotlinDetector;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Helper class for @{@link ScheduledAnnotationBeanPostProcessor} to support reactive cases
+ * without a dependency on optional classes.
+ * @author Simon Baslé
+ * @since 6.1.0 //FIXME
+ */
+abstract class ScheduledAnnotationReactiveSupport {
+
+	static final boolean publisherPresent = ClassUtils.isPresent(
+			"org.reactivestreams.Publisher", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	static final boolean reactorPresent = ClassUtils.isPresent(
+			"reactor.core.publisher.Flux", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	static final boolean coroutinesReactorPresent = ClassUtils.isPresent(
+			"kotlinx.coroutines.reactor.MonoKt", ScheduledAnnotationReactiveSupport.class.getClassLoader());
+
+	/**
+	 * Checks that if the method is reactive, it can be scheduled. Methods are considered
+	 * eligible for reactive scheduling if they either return an instance of
+	 * {@code Publisher} or are a Kotlin Suspending Function. If the method isn't matching
+	 * these criteria then this check returns {@code false}.
+	 * <p>For reactive scheduling, Reactor MUST be present at runtime. In the case of
+	 * Kotlin, the Coroutine {@code kotlinx.coroutines.reactor} bridge MUST also be
+	 * present at runtime (in order to invoke suspending functions as a {@code Mono}).
+	 * Provided that is the case, this method returns {@code true}. Otherwise, it throws
+	 * an {@code IllegalStateException}.
+	 * @throws IllegalStateException if the method is reactive but Reactor and/or the
+	 * Kotlin coroutines bridge are not present at runtime
+	 */
+	static boolean isReactive(Method method) {
+		if (KotlinDetector.isKotlinPresent() && KotlinDetector.isSuspendingFunction(method)) {
+			if (coroutinesReactorPresent) {
+				return true;
+			}
+			else {
+				throw new IllegalStateException("Kotlin suspending functions may only be annotated with @Scheduled"
+						+ "if Reactor and the Reactor-Coroutine bridge (kotlinx.coroutines.reactor) are present at runtime");
+			}
+		}
+		if (!publisherPresent) {
+			return false;
+		}
+		if (!Publisher.class.isAssignableFrom(method.getReturnType())) {
```

Comment:
May trigger a NPE if `Publisher` is not present in the classpath I think.

Reply:
the `if (!publisherPresent)` check above should prevent that part of the code to be reached and thus avoid a class-loading exception